service/raft/raft_group0_client.hh

```diff
@@ -190,4 +191,59 @@ public:
     void set_query_result(utils::UUID query_id, service::broadcast_tables::query_result qr);
 };
 
+using mutations_generator = coroutine::experimental::generator<mutation>;
+
+// mutations_collector is used to gather mutations which are side effects
+// of functions execution. They need to be announced under single guard
+// for atomicity. As functions which produce mutations may embed each other
+// we need to decouple announcing step to a common external place here.
+// It also supports generator callbacks to avoid holding too many mutations
+// in memory.
+//
+// There is no guarantee about ordering of mutations within singe transaction.
+// Single mutations_collector object represents a single transaction.
+// If size or number of mutations is too big for raft too handle it will be
+// rejected.
+class mutations_collector {
+public:
+    using generator_func = std::function<mutations_generator(api::timestamp_type t)>;
+private:
+    std::vector<mutation> _muts;
+    std::vector<generator_func> _generators;
+    std::optional<::service::group0_guard> _guard;
+
+    future<std::vector<mutation>> materialize_mutations();
+public:
+    mutations_collector(::service::group0_guard&& g) : _guard(std::move(g)) {}
+    // Constructor with optional guard used to handle both legacy and current code.
+    // There is no guard for legacy code but the whole class may be passed
+    // through to simplify the flow.
+    mutations_collector(std::optional<::service::group0_guard> g) {
+        if (g) {
+            _guard = std::move(*g);
+        }
+    };
+
+    // Annotation helper for cases where we need collector (e.g. some interface)
+    // but the code is fully legacy and the collector won't be used.
+    static mutations_collector unused() {
+        return mutations_collector(std::nullopt);
+    }
+
+    mutations_collector(mutations_collector&) = delete;
+    mutations_collector(mutations_collector&&) = default;
+
+    // Gets timestamp which should be used when building mutations.
+    api::timestamp_type write_timestamp() const;
+
+    void add_mutation(mutation m);
+    void add_mutations(std::vector<mutation> ms);
+    void add_generator(generator_func f);
+
+    // Commits the data, must be called only once. Nop if there was no guard provided.
+    future<> announce(::service::raft_group0_client& group0_client, std::string_view description, seastar::abort_source& as, std::optional<::service::raft_timeout> timeout);
```

Comment:
If it can be called only once may be it should be marked as '&&'

Reply:
I see only one potential thing if in the future we'd like to extend it to generate more than 1 transaction for whatever reason (there was something lately around auth restore)

but we can revert it then, so I'll mark  now with &&